<a href="https://colab.research.google.com/github/JoshEliell/ColabFiles/blob/main/Actividad_Calificada_07_Datos_Duplicados_274705.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |  Victor Joshua Hernandez Ramon |
| **Matrícula** |  274705 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [76]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [77]:
# Escribe tu código aquí
# Conteo conservando la primera aparicion (filas que sobran)
dup_first = hoteles.duplicated(keep='first').sum()

# Conteo conservando la última aparicion (filas que sobran)
dup_last = hoteles.duplicated(keep='last').sum()

# Conteo de todas las filas que forman parte de un grupo repetido
dup_false = hoteles.duplicated(keep=False).sum()

print(f"Filas sobrantes (keep='first'): {dup_first}")
print(f"Filas sobrantes (keep='last'): {dup_last}")
print(f"Filas involucradas en repeticiones (keep=False): {dup_false}")

Filas sobrantes (keep='first'): 31994
Filas sobrantes (keep='last'): 31994
Filas involucradas en repeticiones (keep=False): 40165


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [78]:
# Escribe tu código aquí
# Porcentaje de filas sobrantes respecto al total del dataset
porcentaje_sobrantes = (dup_first / len(hoteles)) * 100

# Grupos repetidos: la diferencia entre filas involucradas y filas sobrantes
grupos_repetidos = dup_false - dup_first

print(f"Porcentaje de filas sobrantes: {porcentaje_sobrantes:.2f}%")
print(f"Número de grupos repetidos: {grupos_repetidos}")

Porcentaje de filas sobrantes: 26.80%
Número de grupos repetidos: 8171


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [79]:
# Escribe tu código aquí
# Copia para no alterar el original y booleano
hoteles_dup = hoteles.copy()
hoteles_dup['es_duplicado'] = hoteles_dup.duplicated(keep='first')

# Agrupacion por hotel para sumar duplicados
dup_por_hotel = hoteles_dup.groupby('hotel')['es_duplicado'].agg(['sum', 'count'])

# Calculo del porcentaje por hotel
dup_por_hotel['porcentaje (%)'] = (dup_por_hotel['sum'] / dup_por_hotel['count']) * 100
print(dup_por_hotel)

                sum  count  porcentaje (%)
hotel                                     
City Hotel    25902  79330       32.650952
Resort Hotel   6092  40060       15.207189


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

El problema de duplicidad es grande, afectando al 26.80% del dataset original. Es diferente para cada hotel, el City Hotel tiene una proporcion mayor de filas sobrantes frente al Resort Hotel, esto nos dice que hay una diferencia en como se registran las reservas en cada establecimiento.

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [80]:
# Escribe tu código aquí
# Mascara para las filas
mascara_todos = hoteles.duplicated(keep=False)

columnas_vista = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
                  'arrival_date_day_of_month', 'adults', 'country', 'market_segment',
                  'deposit_type', 'adr', 'reservation_status']

# Se ordena por fecha de llegada y lead_time para agrupar visualmente reservas que
# ocurrieron para el mismo dia y se hicieron con la misma anticipacion
print(hoteles[mascara_todos].sort_values(
    by=['arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'lead_time']
)[columnas_vista].head(12))

              hotel  lead_time  arrival_date_year arrival_date_month  \
40574    City Hotel          2               2015             August   
40580    City Hotel          2               2015             August   
40588    City Hotel          2               2015             August   
846    Resort Hotel         10               2015             August   
848    Resort Hotel         10               2015             August   
15943  Resort Hotel         67               2015             August   
15968  Resort Hotel         67               2015             August   
843    Resort Hotel         84               2015             August   
844    Resort Hotel         84               2015             August   
40562    City Hotel        117               2015             August   
40563    City Hotel        117               2015             August   
40564    City Hotel        117               2015             August   

       arrival_date_day_of_month  adults country market_segment

### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [81]:
# Escribe tu código aquí
# Conteo de ocurrencias de cada combinacion de filas (incluyendo NaNs)
tamanos = hoteles.value_counts(dropna=False)

# Cuantos grupos hay de cada tamaño
print("Grupos por tamaño:")
print(tamanos[tamanos > 1].value_counts().head(10))

# Tamaño del grupo mas grande
print(f"\nTamaño del grupo más grande: {tamanos.max()}")
print("Valores del grupo más grande:")
print(tamanos.head(1))

Grupos por tamaño:
count
2     5034
3      969
4      497
5      275
6      171
10     123
7      115
8      103
20      97
9       93
Name: count, dtype: int64

Tamaño del grupo más grande: 180
Valores del grupo más grande:
hotel       is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type  adr    required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date
City Hotel  1            277        2016               November            46                        7                          1                        2                     2       0.0       0       BB    PRT     

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [82]:
# Escribe tu código aquí
duplicados = hoteles[mascara_todos]

print("Distribucion de Market Segment")
print("Total:\n", hoteles['market_segment'].value_counts(normalize=True))
print("Duplicados:\n", duplicados['market_segment'].value_counts(normalize=True))

print("\nDistribucion de Deposit Type")
print("Total:\n", hoteles['deposit_type'].value_counts(normalize=True))
print("Duplicados:\n", duplicados['deposit_type'].value_counts(normalize=True))

print("\nTasa de Cancelacion")
print(f"Total: {hoteles['is_canceled'].mean():.2f}")
print(f"Duplicados: {duplicados['is_canceled'].mean():.2f}")

Distribucion de Market Segment
Total:
 market_segment
Online TA        0.473046
Offline TA/TO    0.202856
Groups           0.165935
Direct           0.105587
Corporate        0.044350
Complementary    0.006223
Aviation         0.001985
Undefined        0.000017
Name: proportion, dtype: float64
Duplicados:
 market_segment
Groups           0.416109
Offline TA/TO    0.303996
Online TA        0.205876
Corporate        0.036699
Direct           0.034981
Complementary    0.001867
Aviation         0.000473
Name: proportion, dtype: float64

Distribucion de Deposit Type
Total:
 deposit_type
No Deposit    0.876464
Non Refund    0.122179
Refundable    0.001357
Name: proportion, dtype: float64
Duplicados:
 deposit_type
No Deposit    0.641678
Non Refund    0.356430
Refundable    0.001892
Name: proportion, dtype: float64

Tasa de Cancelacion
Total: 0.37
Duplicados: 0.58


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

a) Al carecer de un ID unico, es probable que estos no sean errores de captura, sino reservas distintas correspondientes a grupos. Evidencia es que los segmentos de mercado como Groups u Online TA aparecen mucho en los duplicados. Una agencia que reserva 5 habitaciones identicas el mismo dia generara 5 filas identicas.


b) Pandas trata los NaN como valores iguales al utilizar duplicated(). Si varias reservas carecen de agencia asignada, el sistema las agrupa como identicas, esto aumentando el riesgo de eliminar reservas reales que omitieron ese dato.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [83]:
# Escribe tu código aquí
# Eliminar sobrantes conservando la primera y renumerando el indice
hoteles_limpio = hoteles.drop_duplicates(keep='first', ignore_index=True)

print(f"Filas antes: {len(hoteles)}")
print(f"Filas después: {len(hoteles_limpio)}")
print(f"Filas eliminadas: {len(hoteles) - len(hoteles_limpio)}")

Filas antes: 119390
Filas después: 87396
Filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [84]:
# Escribe tu código aquí
# Agrupar y calculo de medias para el dataset original
resumen_original = hoteles.groupby('hotel')[['is_canceled', 'adr']].mean()
resumen_original['reservas'] = hoteles.groupby('hotel').size()
resumen_original = resumen_original.reset_index()

# Agrupación y cálculo de medias para el dataset limpio
resumen_limpio = hoteles_limpio.groupby('hotel')[['is_canceled', 'adr']].mean()
resumen_limpio['reservas'] = hoteles_limpio.groupby('hotel').size()
resumen_limpio = resumen_limpio.reset_index()

### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [85]:
# Escribe tu código aquí
# Asignación de versión y concatenación
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'

comparacion = pd.concat([resumen_original, resumen_limpio], ignore_index=True)
comparacion = comparacion.sort_values(by='hotel')
print(comparacion)

          hotel  is_canceled         adr  reservas         version
0    City Hotel     0.417270  105.304465     79330        original
2    City Hotel     0.300386  110.985944     53428  sin duplicados
1  Resort Hotel     0.277634   94.952930     40060        original
3  Resort Hotel     0.234809   99.025346     33968  sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

a) La tasa de cancelacion es la que mas cambia, especialmente en el City Hotel. Debido a que eliminamos mayoritariamente reservas grupales identicas. Si un grupo grande no cancelo, sus varios registros identicos reducian la tasa de cancelación general.


b) En un modelo predictivo, dejar los duplicados provocaria data leakage y overfitting. Si una fila de un grupo identico entra en el conjunto de entrenamiento y su copia en el de prueba, el modelo tomara de manera muy optimista esto, ya que memorizara los casos exactos en lugar de generalizar.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [86]:
# Escribe tu código aquí
# Verificar la eliminacion de duplicados exactos
print("Duplicados restantes:", hoteles_limpio.duplicated().sum())

# Verificacipn del indice
print("Indices repetidos:", hoteles_limpio.index.duplicated().sum())
print("Indice maximo correcto:", hoteles_limpio.index.max() == (len(hoteles_limpio) - 1))

Duplicados restantes: 0
Indices repetidos: 0
Indice maximo correcto: True


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [87]:
# Escribe tu código aquí
# Todas las columnas excepto la fecha de ultimo estado
columnas_sin_fecha = hoteles_limpio.columns.drop('reservation_status_date')

# Checar duplicados parciales
parciales = hoteles_limpio.duplicated(subset=columnas_sin_fecha, keep=False)
print(f"1. Filas sobrantes parciales: {hoteles_limpio.duplicated(subset=columnas_sin_fecha).sum()}")

# Visualizar
columnas_vista_parcial = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
                          'arrival_date_day_of_month', 'adr', 'reservation_status',
                          'reservation_status_date']
print('\n 2.')
print(hoteles_limpio[parciales].sort_values(by=['arrival_date_year', 'lead_time'])[columnas_vista_parcial].head(10))

# Distribucion del estado (filas involucradas)
print('\n 3.')
print(hoteles_limpio[parciales]['reservation_status'].value_counts())

1. Filas sobrantes parciales: 245

 2.
              hotel  lead_time  arrival_date_year arrival_date_month  \
14716  Resort Hotel          8               2015           December   
14717  Resort Hotel          8               2015           December   
35572    City Hotel          9               2015          September   
35574    City Hotel          9               2015          September   
54170    City Hotel         14               2015            October   
54474    City Hotel         14               2015            October   
55781    City Hotel         24               2015           December   
56010    City Hotel         24               2015           December   
55315    City Hotel         32               2015           November   
55682    City Hotel         32               2015           November   

       arrival_date_day_of_month    adr reservation_status  \
14716                          5   68.0           Canceled   
14717                          5   68.0     

**Respuesta 4.2:**

Probablemente representan reservas grupales donde las personas hicieron el check-out en dias distintos o sus cancelaciones se procesaron en fechas diferentes.

No se deben eliminar. Ya que estos son datos reales del proceso del hotel y representan entidades diferentes comportandose de manera diferente en cierta parte final del proceso.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

Criterio: Se evaluo la coincidencia exacta en las 32 variables del dataset.   

Regla de conservación: Se utilizó keep='first' para preservar un registro base de cada bloque de reservas identicas, esto asumiendo que al menos un registro es real y tiene la informacion original de la reserva.

Resultado: Se eliminaron casi 32,000 registros, reduciendo el numero de registros y ajustando las tasas de cancelacion que aumentaban.

Supuestos y limitaciones: Se asume que las filas idénticas no son clientes distintos por accidente, hay una limitacion impuesta por la falta de un reservation_id o room_number. Existe un alto chance de haber eliminado reservas grupales legitimas que compartían las mismas condiciones.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

1- No convendría eliminarlos si el objetivo del analisis es la planificacion operativa y logística del hotel. Dado que el dataset representa la demanda de reservas, esas filas identicas probablemente corresponden a habitaciones reservadas por un mismo grupo o agencia. Al eliminarlas, estaríamos subestimando el volumen real de huéspedes que llegan al hotel.

2- Solicitaria un ID de reserva y el nmero de habitacion. Esto permitiria diferenciar facilmente si 5 filas iguales son un error de sistema guardando la misma transacción 5 veces, o si son las 5 habitaciones juntas de una misma familia.

3- Un conteo de 0 en duplicated().sum() solo indica que no hay filas idénticas carácter por carácter. El dataset puede seguir lleno de duplicados reales ocultas por diferencias de formato, o marcas de tiempo diferentes generadas automáticamente por el sistema, como se vio con reservation_status_date.  

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.